### PR TITLE
Issue #11655: allow suppressing Google style warnings with annotations and comments

### LINKED ISSUE #11655: Update google_checks.xml to have the SuppressionCommentFilter and SuppressWarningsHolder modules in the config by default (and by extension, SuppressWarningsFilter)
I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

**How it works Now:**

```bash
/var/tmp $ javac MyClass.java
#[[MAKE SURE THERE IS SUCCESSFUL COMPILATION]]

/var/tmp $ cat google_checks.xml
...
<module name="Checker">
  <!-- IMPORTANT: Allows use of SuppressWarnings annotation -->
  <module name="SuppressWarningsFilter"/>
  ...
  <module name="TreeWalker">
    <module name="SuppressWarningsHolder" />
    <module name="SuppressionCommentFilter">
      <property name="offCommentFormat" value="CHECKSTYLE.OFF\: ([\w\|]+)" />
      <property name="onCommentFormat" value="CHECKSTYLE.ON\: ([\w\|]+)" />
      <property name="checkFormat" value="$1" />
    </module>
...
</module>


/var/tmp $ cat MyClass.java
class MyClass {
  public void aMethod() {
  }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.2-all.jar -c google_checks.xml MyClass.java
Starting audit...
[WARN] /var/tmp/MyClass.java:2:15: Method name 'aMethod' must match pattern '^[a-z][a-z0-9][a-zA-Z0-9_]*$'. [MethodName]
Audit done.
```

After using `@SuppressWarnings` 

```
/var/tmp $ cat MyClass.java
class MyClass {
  @SuppressWarnings("checkstyle:methodname")
  public void aMethod() {
  }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.2-all.jar -c google_checks.xml MyClass.java
Starting audit...
Audit done.
```

Using special comments to switch checkstyle off and on:

```
/var/tmp $ cat MyClass.java
class MyClass {
  // CHECKSTYLE.OFF: MethodName
  public void aMethod() {
  }
  // CHECKSTYLE.ON: MethodName
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.2-all.jar -c google_checks.xml MyClass.java
Starting audit...
Audit done.
```

**Is your feature request related to a problem? Please describe.**
No, this is just additional config in `google_checks.xml`.

**Describe the solution you'd like**
Add the ability to suppress style checks via annotations and special comments around the code. This will offer users alternatives to suppress checks other than using XML.

**Additional context**
Some of these changes have been already adopted by Google https://github.com/GoogleCloudPlatform/java-repo-tools/blob/main/third_party/checkstyle/google_checks.xml

**References**
* http://stackoverflow.com/a/22556386/101923
* http://stackoverflow.com/questions/4023185

### PR DESCRIPTION
**Pull request by ddcprg**
Fixes #11655

### CODE REVIEW

File: src/main/resources/google_checks.xml
Comment (ddcprg) [inline src/main/resources/google_checks.xml:119]: `LITERAL_YIELD` would raise me an error that says the token is not recognised

Comment (ddcprg) [reply src/main/resources/google_checks.xml:119]: Looks like this change has broken the unit tests. I didn't look into why it would fail. I'll push it until I find out

**Comment by nick-mancuso** [comment]
@ddcprg issue is approved, please see https://github.com/checkstyle/checkstyle/issues/11655#issuecomment-1132928670

**Comment by romani** [comment]
@ddcprg, ping

**Comment by ddcprg** [comment]
@romani I have time to look at this later today

File: src/main/resources/google_checks.xml
Comment (ddcprg) [inline src/main/resources/google_checks.xml:61]: a different comment format may be more suitable to avoid confusion with `SuppressionCommentFilter`

File: src/main/resources/google_checks.xml
Comment (ddcprg) [inline src/main/resources/google_checks.xml:59]: I'll remove this comment

File: src/main/resources/google_checks.xml
Comment (nick-mancuso) [inline src/main/resources/google_checks.xml:63]: Please explain why we do not leave properties as default value (except for comment format)

**Review by nick-mancuso — CHANGES_REQUESTED** [review]
Items:

File: src/main/resources/google_checks.xml
Comment (ddcprg) [reply src/main/resources/google_checks.xml:63]: done

**Comment by ddcprg** [comment]
@nick-mancuso changes made, please feel free to review again any time

**Comment by nick-mancuso** [comment]
Github, generate site

**Review by nick-mancuso — CHANGES_REQUESTED** [review]
We do not have a testing structure in place to test config with suppressions for google style (not that I can find, anyway). Please show CLI output with your new config file and working example with suppressions in place. We can use parts of your example to address the below item:

Item:

File: src/main/resources/google_checks.xml
Comment (romani) [inline src/main/resources/google_checks.xml:53]: Please put filters next/close to 
`<module name="SuppressionXpathFilter`

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/main/resources/google_checks.xml
Comment (romani) [inline src/main/resources/google_checks.xml:19]: Please move formatting to separate commit "minor: formatting" keep it in same PR.

File: src/main/resources/google_checks.xml
Comment (romani) [inline src/main/resources/google_checks.xml:20]: I am not sure this comment is required. 
User need to read doc for module to understand what is it for.

Or we need to put this description to our xdoc documentation

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by romani** [comment]
@ddcprg, ping

File: src/xdocs/google_style.xml
Comment (ddcprg) [reply src/xdocs/google_style.xml]: addressed, I'll wait until the build generates the docs to double check

File: src/main/resources/google_checks.xml
Comment (ddcprg) [reply src/main/resources/google_checks.xml:19]: moved to a separate commit as per your suggestion

File: src/main/resources/google_checks.xml
Comment (ddcprg) [reply src/main/resources/google_checks.xml]: I've removed the comment

**Comment by ddcprg** [comment]
@romani I can't find the link to the generated site, would you mind telling me where to get it from?

**Comment by ddcprg** [comment]
thank you @Vyom-Yadav @romani @nick-mancuso let me know if the docs look good enough https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/1d44145_2022082908/google_style.html

File: src/xdocs/google_style.xml
Comment (romani) [inline src/xdocs/google_style.xml:2317]: Please remove SuppressWarningsHolder, it is very hacky module that we are not very proud of. It is helper module, it is not a filter.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/xdocs/google_style.xml
Comment (romani) [inline src/xdocs/google_style.xml:2331]: You meant SuppressionCommentFilter
And SuppressWithNearbyCommentFilter is for next block.

File: src/xdocs/google_style.xml
Comment (romani) [inline src/xdocs/google_style.xml:2341]: Please make filers as link to their documentation
.

File: src/xdocs/google_style.xml
Comment (romani) [inline src/xdocs/google_style.xml:2343]: Better to wrap value of annotation with SuppressWarnings

File: src/xdocs/google_style.xml
Comment (romani) [inline src/xdocs/google_style.xml:2337]: Better to show how to activate validation back , in other case  whole file below this OFF will be suppressed.

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by ddcprg** [comment]
Github, generate site

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by ddcprg** [comment]
@romani I still can't merge the PR, if you can please press the button

**Comment by romani** [comment]
Only maintainers can merge.
You need to pass review. PR is assigned for next reviewer.

**Review by nick-mancuso — APPROVED** [review]
@ddcprg thanks a lot for making this happen!

**Pull request merged**
